platform/pwl.c

```diff
@@ -1393,10 +1417,12 @@ pwl_display_destroy (PwlDisplay *self)
         g_clear_pointer (&self->application_id, g_free);
         g_clear_pointer (&self->window_title, g_free);
 
-        g_clear_pointer (&self->xkb_data.state, xkb_state_unref);
-        g_clear_pointer (&self->xkb_data.compose_state, xkb_compose_state_unref);
-        g_clear_pointer (&self->xkb_data.keymap, xkb_keymap_unref);
-        g_clear_pointer (&self->xkb_data.context, xkb_context_unref);
+        wl_list_for_each_safe (item, tmp, &self->seats, link) {
+            g_clear_pointer (&item->xkb_data.state, xkb_state_unref);
+            g_clear_pointer (&item->xkb_data.compose_state, xkb_compose_state_unref);
+            g_clear_pointer (&item->xkb_data.keymap, xkb_keymap_unref);
+            g_clear_pointer (&item->xkb_data.context, xkb_context_unref);
+        }
```

Comment:
You have to unlink each seat item from the list, and free the memory used by the corresponding `PwlSeat` value. You may want to add a `pwl_seat_destroy()` function which takes care of destroying/unreffing the members of `PwlSeat` and the deallocates the `struct` itself. That way the only work to do here is iterate the `wl_list` unlinking elements and calling `pwl_seat_destroy()` on them.